src/electron/routing_service.ts

```diff
@@ -251,52 +291,68 @@ function installWindowsRoutingServices(): Promise<void> {
   return executeCommandAsRoot(script);
 }
 
-async function installLinuxRoutingServices(): Promise<void> {
-  const OUTLINE_PROXY_CONTROLLER_PATH = path.join('tools', 'outline_proxy_controller', 'dist');
-  const LINUX_INSTALLER_FILENAME = 'install_linux_service.sh';
-  // [InstallationFileName, isExecutable]...
-  const LINUX_SERVICE_FILE_NAMES: Array<[string, boolean]> = [
-    [LINUX_INSTALLER_FILENAME, true],
-    ['OutlineProxyController', true],
-    ['outline_proxy_controller.service', false],
-  ];
-
-  // These Linux service files are located in a mounted folder of the AppImage, typically
-  // located at /tmp/.mount_Outlinxxxxxx/resources/. These files can only be acceeded by
-  // the user who launched Outline.AppImage, so even root cannot access the files or folders.
-  // Therefore we have to copy these files to a normal temporary folder, and execute them
-  // as root.
-  //
-  // Also, we are copying individual files instead of the entire folder because they are in
-  // electron's "asar" archive (which is returned by getAppPath). Electron tries to inject
-  // some logic to node's fs API and make sure the caller can access files in the archive.
-  // However directories are not supported.
-  //
-  // References:
-  // - https://github.com/AppImage/AppImageKit/issues/146
-  // - https://xwartz.gitbooks.io/electron-gitbook/content/en/tutorial/application-packaging.html
-  const tmp = await fsextra.mkdtemp('/tmp/');
-  const srcFolderPath = path.join(getAppPath(), OUTLINE_PROXY_CONTROLLER_PATH);
-
-  console.log(`copying service installation files to ${tmp}`);
-  for (const [filename, executable] of LINUX_SERVICE_FILE_NAMES) {
-    const dest = path.join(tmp, filename);
-    await fsextra.copy(path.join(srcFolderPath, filename), dest, {overwrite: true});
-    if (executable) {
-      await fsextra.chmod(dest, 0o755);
-    }
+/**
+ * The actual steps to install a Linux Outline service. The caller should already
+ * granted the admin permission to the process before calling this method.
+ */
+export async function doInstallLinuxRoutingServices(): Promise<void> {
+  if (!isLinux) {
+    throw new Error('please call this method in Linux');
   }
-  console.log(`all service installation files copied to ${tmp} successfully`);
 
-  await executeCommandAsRoot(path.join(tmp, LINUX_INSTALLER_FILENAME));
```

Comment:
I think we can make this a minimal PR by simply calling `chattr -i` here, and checking the checksum before calling execute.

Reply:
Yes, then we will need three checksums (because we are copying three files out). I can make the change in this PR to unblock security review. And I will refine it by putting all logic in the single executable file.